### DISCUSSION TITLE
Styling Font Color and Background Color always reset (removed) when I reopen Editor Content Tiptap

### DISCUSSION THREAD

**Discussion by Arik14** [ueberdosis/tiptap#1743] [General Questions]
**Description**
I have Text Editor component which consist of `<editor-content>` from Tiptap inside it. This component only created when I click the text.
I add feature to edit font color. I look at the documentation an found the instruction https://www.tiptap.dev/experiments/font-color/#commands. But when I try to install the package, it's failed. I try to look at the Tiptap github, The package is not existed (https://github.com/ueberdosis/tiptap/blob/main/packages/extension-font-color/).
So I try to create custom extension for editing font color by myself.

**FontSize.js**
```
import { Mark, mergeAttributes } from "@tiptap/core";

const FontColor = Mark.create({
	name: "fontColor",
	defaultOptions: {
		HTMLAttributes: {},
		defaultFontColor: "#333333"
	},
	addAttributes() {
		return {
			fontColor: {
				default: null,
				parseHTML: element => ({
					fontColor:
						element.style.color || this.options.defaultFontColor
				}),
				renderHTML: attributes => ({
					style: `color: ${attributes.fontColor}`
				})
			}
		};
	},
	parseHTML() {
		return [
			{
				tag: "span"
			}
		];
	},
	renderHTML({ HTMLAttributes }) {
		return [
			"span",
			mergeAttributes(this.options.HTMLAttributes, HTMLAttributes),
			0
		];
	},
	addCommands() {
		return {
			setFontColor: fontColor => ({ commands }) => {
				return commands.setMark("fontColor", { fontColor });
			}
		};
	}
});

export default FontColor;
```

And also I create custom extension for Font Size too.
**FontSize.js**
```
import { Mark, mergeAttributes } from "@tiptap/core";

const FontSize = Mark.create({
	name: "fontSize",
	defaultOptions: {
		HTMLAttributes: {},
		defaultFontSize: "16px"
	},
	addAttributes() {
		return {
			fontSize: {
				default: null,
				parseHTML: element => ({
					fontSize:
						element.style.fontSize || this.options.defaultFontSize
				}),
				renderHTML: attributes => ({
					style: `font-size: ${attributes.fontSize}`
				})
			}
		};
	},
	parseHTML() {
		return [
			{
				tag: "span"
			}
		];
	},
	renderHTML({ HTMLAttributes }) {
		return [
			"span",
			mergeAttributes(this.options.HTMLAttributes, HTMLAttributes),
			0
		];
	},
	addCommands() {
		return {
			setFontSize: fontSize => ({ commands }) => {
				return commands.setMark("fontSize", { fontSize });
			}
		};
	}
});

export default FontSize;
```

Suppose I have this content.
```
<p>
  Hello World
</p>
```
Then I select the text `World` and apply font color with this command
```
this.editor.chain().focus().setFontColor(value).run();
```
where the value is string rgba, for example `rgba(255,255,0,1)`.
It succeed. I look at the html, it's already changed into this.
```
<p>
  Hello <span style="color: rgba(255, 255, 0, 1)">World</span>
</p>
```
Then, I close the editor which this trigger editor content to destroy.
The problem is, when I reopen the Text Editor, which basically iniate the editor content again, it seem remove color styling.
I checked at the html, it became just like this.
```
<p>
  Hello 
  <span style="font-size: 16px">World</span>
</p>
```

Why is the color styling is disappear?

I want the styling is coexist with other styling (if possible), 
```
<p>
  Hello 
  <span style="font-size: 16px; color: rgba(255, 255, 0, 1)">World</span>
</p>
```
or to be nested element (this is also okay).
```
<p>
  Hello 
  <span style="font-size: 16px">
     <span style="color: rgba(255, 255, 0, 1)">World</span>
  </span>
</p>
```

I already try the custom extension from this https://gist.github.com/sabetAI/e7ac09da82ea4e1e01929beb836ffa41
But, unfortunately, it couldn't event change the style color at all.

I try replicate in this sandbox in here https://codesandbox.io/s/sweet-fast-tlzf7?file=/src/components/CustomEditor.vue (It over simplified version of my project). But it worked, that's weird.
(You can click button Toggle Editor, select text Hello and click button setColor. Then click Toggle Editor twice to reopen component Custom Editor. The text Hello is still red, which is worked).
The principle in my real project is quite the same. There is inline Editor for Tiptap which only appear when the text is clicked.
But in my case, the color is always replaced by font size styling.
You can watch my demo in here. I also show the element below so you can see what is changing in html. https://www.loom.com/share/e9d7cdec386749d3a52be7e04820236e
I don't know what's wrong with the code in my real project.

Please kindly help me.
Thank you.

Arik.

**Comment by hanspagel** [comment]
I just opened a PR for the official extension, expect it to be released this week.

https://github.com/ueberdosis/tiptap/pull/1744

**Reply by Arik14** [reply]
Oh, thank you for that. It will help not only me, but also many others too.
I will update my code as soon as the extension is released. Hopefully the merging process is smooth.

***
But will the extension create something like this?
```
<p>
  Hello 
  <span style="font-size: 16px">
     <span style="color: rgba(255, 255, 0, 1)">World</span>
  </span>
</p>
```

I also already tried Highlight extension. It uses <mark> rather than <span>. And it possible to be nested
```
<span style="font-size: 16px">
  <mark data-color="..." style="background-color: ..."></mark>
</span>
```
I am still wondering why in my case, the <span> with color styling is replaced by <span> with font size styling, instead of nested (making it as child). Is the same element not allowed to be nested?
Did I missed something here?

**Reply by Arik14** [reply]
Thank you, it worked. I forget to add TextStyle in extension instance Editor, and also other reason (explain in reply to philippkuehn). But it worked anyway.

I will mark this as answer. (But encounter problem when click Mark as Answer. I don't if it problem with my browser, or bug in github. I'll do it again later).

**Comment by Arik14** [comment]
Hello, sorry for keep asking question. 
Today, I finally could reproduce the problem in sandbox, here is the link https://codesandbox.io/s/sweet-fast-tlzf7?file=/src/components/HelloWorld.vue.

Before I click button `toggleEditor`, as you can see the html string contain span with color style.
![Screenshot from 2021-08-19 09-27-20](https://user-images.githubusercontent.com/54788602/129998182-40f49e66-0336-44e4-b7ad-3360c1acdffe.png)

And after I click button toggleEditor, the span with color style is disappeared and replaced by span with font-size style.
![Screenshot from 2021-08-19 09-28-13](https://user-images.githubusercontent.com/54788602/129998222-5c7dab1e-0088-477d-bad5-08277db64489.png)

I think this it is somehow related with this part of code
```
extensions: [StarterKit, FontSize, FontColor],
```
![Screenshot from 2021-08-19 09-34-48](https://user-images.githubusercontent.com/54788602/129998638-e0bb8ee7-9927-4274-a21b-3625d3600a6e.png)

It seems like order in the array is related to the priority. In this case, Font Size will override the span with Font Color.
If I flip the order, the Font Color will override any span with Font Size.
```
extensions: [StarterKit, FontColor, FontSize],
```
Here is before I click toggle Editor button.
![Screenshot from 2021-08-19 09-23-16](https://user-images.githubusercontent.com/54788602/129998806-362ea2f9-9b36-4106-8b74-7797b21212d1.png)

Here is after I click toggle Editor button.
![Screenshot from 2021-08-19 09-23-54](https://user-images.githubusercontent.com/54788602/129998841-1b9361c0-9625-44df-9099-00584d4b09b1.png)

What should I to maintain the nested span like this?
```
<p>
  Hello 
  <span style="font-size: 16px">
     <span style="color: rgba(255, 255, 0, 1)">World</span>
  </span>
</p>
```

**Reply by philippkuehn** [reply]
There is an official [TextStyle](https://www.tiptap.dev/api/marks/text-style) extension that can render a single `<span>`. Other extensions can extend it (like [FontFamily](https://www.tiptap.dev/api/extensions/font-family) or the new [Color](https://github.com/ueberdosis/tiptap/pull/1744) extension).

**Reply by Arik14** [reply]
Oh, the PR is already merged.
I will try again.

But yesterday, I try to follow the code from extension-color (when its status is still PR). Extend the Text Style.
https://github.com/ueberdosis/tiptap/blob/c9c7109cd766ea7f6064d9d432c953427a762d88/packages/extension-color/src/color.ts
But it's not worked, Perhaps, I made mistake while rewrite it in JS.
(I may showing sandbox of my code when I'm home).

**Reply by Arik14** [reply]
Okay, thank you. It's worked.
And I just found the problem in my case too.
It was my custom extension FontSize. I use Mark.create instead of extend from TextStyle like in FontFamily.
This make whatever styling always replace by span with font-size.

Here is the sandbox https://codesandbox.io/s/sweet-fast-tlzf7?file=/src/components/FontSize.js
As you can see, I use the new Color extension together with FontSize.js (my custom extension).
When the button `toggle editor` is clicked, the span with color style is replaced by span with font-size.

So the problem is not with the Color. This explains why my previous attempt to set color with my own custom extension is still worked. But replaced with font size whenever the editor is reopen. I never checked the Font Size implementation.
Sorry for that.

Thank you for your help.
Hopefully new Tiptap will grow larger and be far better than before.
